keyboards/1upkeyboards/pi40/pi40.c

```diff
@@ -0,0 +1,306 @@
+    };
+
+#endif
+
+#ifdef RGB_MATRIX_ENABLE
+led_config_t g_led_config = { {
+    // Key Matrix to LED Index
+    {  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11 },
+    { 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12 },
+    { 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35 },
+    { 46, 45, 44, 43, 42, 41, NO_LED, 40, 39, 38, 37, 36 }
+}, {
+    // LED Index to Physical Position
+    {   9,  8 }, {  28,  8 }, {  46,  8 }, {  65,  8 }, {  84,  8 }, { 102,  8 }, { 121,  8 }, { 140,  8 }, { 159,  8 }, { 177,  8 }, { 196,  8 }, { 215,  8 },
+    { 215, 24 }, { 196, 24 }, { 177, 24 }, { 159, 24 }, { 140, 24 }, { 121, 24 }, { 102, 24 }, {  84, 24 }, {  65, 24 }, {  46, 24 }, {  28, 24 }, {   9, 24 },
+    {   9, 40 }, {  28, 40 }, {  46, 40 }, {  65, 40 }, {  84, 40 }, { 102, 40 }, { 121, 40 }, { 140, 40 }, { 159, 40 }, { 177, 40 }, { 196, 40 }, { 215, 40 },
+    { 215, 56 }, { 196, 56 }, { 177, 56 }, { 159, 56 }, { 140, 56 }, { 112, 56 }, {  84, 56 }, {  65, 56 }, {  46, 56 }, {  28, 56 }, {   9, 56 }
+}, {
+    // LED Index to Flag
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+} };
+#endif
+
+#ifdef OLED_ENABLE
+
+void init_timer(void){
+   oled_logo_timer = timer_read32();
+};
+
+void user_oled_magic(void) {
+    // Host Keyboard Layer Status
+    oled_write_P(PSTR("Layer: "), false);
+
+    switch (get_highest_layer(layer_state)) {
+        case _ONE:
+            oled_write_P(PSTR("One\n"), false);
+            break;
+        case _TWO:
+            oled_write_P(PSTR("Two\n"), false);
+            break;
+        case _THREE:
+            oled_write_P(PSTR("Three\n"), false);
+            break;
+        case _FOUR:
+            oled_write_P(PSTR("Four\n"), false);
+            break;
+        case _FIVE:
+            oled_write_P(PSTR("Five\n"), false);
+            break;
+        case _SIX:
+            oled_write_P(PSTR("Six\n"), false);
+            break;
+        case _SEVEN:
+            oled_write_P(PSTR("Seven\n"), false);
+            break;
+        case _EIGHT:
+            oled_write_P(PSTR("Eight\n"), false);
+            break;
+        case _NINE:
+            oled_write_P(PSTR("Nine\n"), false);
+            break;
+        case _TEN:
+            oled_write_P(PSTR("Ten\n"), false);
+            break;
+        default:
+            // Or use the write_ln shortcut over adding '\n' to the end of your string
+            oled_write_ln_P(PSTR("Undefined"), false);
+    }
+
+    // Host Keyboard LED Status
+    led_t led_state = host_keyboard_led_state();
+    oled_write_P(led_state.caps_lock ? PSTR("Cap(x) ") : PSTR("Cap( ) "), false);
+    oled_write_P(led_state.num_lock ? PSTR("Num(x) ") : PSTR("Num( ) "), false);
+    oled_write_P(led_state.scroll_lock ? PSTR("Scrl(x)") : PSTR("Scrl( )"), false);
+
+
+    switch (rgblight_get_mode()) {
+        case 1:
```

Comment:
It might be "better" (or at least more accurate) to do something like:
```c
#ifdef ENABLE_RGB_MATRIX_ALPHAS_MODS
        case RGB_MATRIX_ALPHAS_MODS:
            oled_write_ln_P(PSTR("Alphas Mods"), false);
            break;
#endif
```
This way, it renders which modes based on which are enabled, and users could disable modes, if they were so inclined and it wouldn't screw up the oled display.


(and add an `oled_advance_page(true);` after the case to add that blank line).

Reply:
I will have to play around with this on my own for the future, but since we are going to be shipping in a few days I at least want to get this code out now and refine it later.